From 05b0e992de71d24f56cf24e86319fad4c1c5788b Mon Sep 17 00:00:00 2001
From: Nikita Yushchenko <nikita.yoush@cogentembedded.com>
Date: Mon, 30 Sep 2024 20:50:13 +0500
Subject: [PATCH 154/157] net: ethernet: renesas: rswitch: support explicit BAT
 placement

Chains used by vport running in Xen DomU are processed with non-default
OSID. This non-default OSID is also used when loading BAT entry for
the chain - which leads to requirement to have iommu setup that
resolves BAT address with any OSID.

To provide that, have to:
- preallocate memory for BAT as a global reservation in xen.dtb,
- map that reservation to the same address for all domains,
- instruct rswitch driver to place BAT at that address.

This patch adds support for configuring a fixed BAT address in rswitch
device tree node.

Signed-off-by: Nikita Yushchenko <nikita.yoush@cogentembedded.com>

diff --git a/drivers/net/ethernet/renesas/rswitch.c b/drivers/net/ethernet/renesas/rswitch.c
index a28d84bcc816..1da52d47707b 100644
--- a/drivers/net/ethernet/renesas/rswitch.c
+++ b/drivers/net/ethernet/renesas/rswitch.c
@@ -1173,6 +1173,7 @@ struct rswitch_private {
 	struct rswitch_desc *desc_bat;
 	dma_addr_t desc_bat_dma;
 	u32 desc_bat_size;
+	bool explicit_bat_address;
 
 	struct list_head port_list;
 
@@ -2821,8 +2822,12 @@ static int rswitch_desc_alloc(struct rswitch_private *priv)
 
 	priv->desc_bat_size =
 			sizeof(struct rswitch_desc) * RSWITCH_MAX_NUM_CHAINS;
-	priv->desc_bat = dma_alloc_coherent(dev, priv->desc_bat_size,
-					    &priv->desc_bat_dma, GFP_KERNEL);
+	if (priv->explicit_bat_address)
+		priv->desc_bat = memremap(priv->desc_bat_dma,
+					priv->desc_bat_size, MEMREMAP_WT);
+	else
+		priv->desc_bat = dma_alloc_coherent(dev, priv->desc_bat_size,
+					&priv->desc_bat_dma, GFP_KERNEL);
 	if (!priv->desc_bat)
 		return -ENOMEM;
 	for (i = 0; i < RSWITCH_MAX_NUM_CHAINS; i++)
@@ -2833,10 +2838,14 @@ static int rswitch_desc_alloc(struct rswitch_private *priv)
 
 static void rswitch_desc_free(struct rswitch_private *priv)
 {
-	if (priv->desc_bat)
-		dma_free_coherent(&priv->pdev->dev, priv->desc_bat_size,
-				  priv->desc_bat, priv->desc_bat_dma);
-	priv->desc_bat = NULL;
+	if (priv->desc_bat) {
+		if (priv->explicit_bat_address)
+			memunmap(priv->desc_bat);
+		else
+			dma_free_coherent(&priv->pdev->dev, priv->desc_bat_size,
+					priv->desc_bat, priv->desc_bat_dma);
+		priv->desc_bat = NULL;
+	}
 }
 
 static int rswitch_gwca_get(struct rswitch_private *priv)
@@ -5409,6 +5418,8 @@ static int rswitch_probe(struct platform_device *pdev)
 	struct resource *res, *res_serdes;
 	struct device_node *ports_dn, *port_dn;
 	struct rswitch_device *rdev;
+	const void *bat_address_prop;
+	int bat_address_prop_len;
 	int ret;
 
 	res = platform_get_resource(pdev, IORESOURCE_MEM, 0);
@@ -5498,6 +5509,18 @@ static int rswitch_probe(struct platform_device *pdev)
 	if (list_empty(&priv->port_list))
 		dev_warn(&pdev->dev, "could not initialize any ports\n");
 
+	bat_address_prop = of_get_property(pdev->dev.of_node, "bat-address",
+			&bat_address_prop_len);
+	if (bat_address_prop) {
+		if (bat_address_prop_len != sizeof(dma_addr_t))
+			dev_warn(&pdev->dev, "wrong bat-address length\n");
+		else {
+			priv->explicit_bat_address = true;
+			priv->desc_bat_dma = of_read_number(bat_address_prop,
+					bat_address_prop_len / sizeof(u32));
+		}
+	}
+
 	INIT_LIST_HEAD(&priv->vport_list);
 	priv->vport_shared_rx_chain_index = -1;
 
-- 
2.25.1

